dlme_airflow/drivers/json.py

```diff
@@ -0,0 +1,148 @@
+            title:
+              path: "title"
+            thumbnail:
+              path: "resources[0].image"
+
+    and the collection URL returns JSON that looks like:
+
+    {
+      "content": {
+        "results": [
+          {
+            "id": "http://www.loc.gov/item/2017498321/",
+            "title": "[Majmuʻah, or, Collection]",
+            "resources": [
+              "image": "https://tile.loc.gov/image-services/iiif/service:amed:plmp:m154:0334/full/pct:6.25/0/default.jpg"
+            ]
+          },
+          {
+            "id": "http://www.loc.gov/item/00313408/",
+            "title": "Sakīnat al-fuz̤alāʼ mawsūm bi-ism-i tārīkhī-i Bahār-i Afghānī",
+            "resources": [
+              "image": "https://tile.loc.gov/image-services/iiif/service:amed:amedpllc:00406537456:0055/full/pct:6.25/0/default.jpg"
+            ]
+          }
+        ]
+      }
+    }
+
+    You would end up with a CSV with three columns (id, title and thumbnail) and two rows.
+    """
+
+    container = "dataframe"
+    name = "custom_json"
+    version = "0.0.1"
+    partition_access = True
+
+    def __init__(self, collection_url, metadata={}, csv_kwargs={}):
+        super(JsonSource, self).__init__(metadata=metadata)
+        self.collection_url = collection_url
+        self.metadata = metadata
+        self.csv_kwargs = csv_kwargs
+        self.record_limit = self.metadata.get("record_limit", None)
+        self.record_count = 0
+        self._setup_json_paths()
+
+    def read(self):
+        self._load_metadata()
+        df = self._get_dataframe()
+        if self.record_limit:
+            return df.head(self.record_limit)
+        else:
+            return df
+
+    def _get_schema(self):
+        return Schema(
+            datashape=None,
+            dtype=self.csv_kwargs.get("dtype"),
+            shape=None,
+            npartitions=1,
+        )
+
+    def _get_dataframe(self):
+        resp = requests.get(self.collection_url)
+        if resp.status_code != 200:
+            raise Exception(
+                f"HTTP request for {self.collection_url} resulted in {resp.status_code}"
+            )
+        return self._process_json(resp.json())
+
+    def _process_json(self, data):
+        result = self.record_selector.find(data)
+        if len(result) == 0:
+            raise Exception(
+                f"Couldn't find records selector: {self.record_selector.expression}"
+            )
+
+        objects = []
+        for rec in result[0].value:
+            obj = {}
+            for field in self.field_paths:
+                path = field["path"]
+                name = field["name"]
+                result = path.find(rec)
+                if len(result) == 1:
+                    obj[name] = result[0].value
+                elif len(result) > 1:
+                    obj[name] = [m.value for m in result]
+                elif not field["optional"]:
+                    raise Exception(f"{name} is not optional")
```

Comment:
i wouldn't necessarily change it in this PR, but it seems like we should ticket making behavior consistent across drivers when a required field is missing.  e.g. in the IIIF and XML drivers, i think a warning is logged when a required field is missing, but no error is raised.  i don't know which approach would be more useful IRL.

Reply:
Yes, you're right. Let me make it a warning for now at least so it's consistent. Then we can change across the board if it becomes an issue. I wanted to throw exceptions during development because I wanted things to break loudly as I was working on the logic.

Reply:
I converted them to warnings for the moment, and also created this issue: https://github.com/sul-dlss/dlme-airflow/issues/236

I think it's worth discussing what we want to happen here. It seems like logs will not be consulted very often, so the warnings will be largely missed?